### ISSUE TITLE
Multi-transport race condition on passive side

### ISSUE THREAD

**Issue by jrw972** [OpenDDS/OpenDDS#4269]
OpenDDS allows the user to create a transport config that can reference multiple transport instances.  The transport instances are tried in order with the intended behavior that the participant can fail over to lower-priority transport instances.  A link between a datawriter and datareader is called a DataLink.

In the transport framework, datawriters are active and datareaders are passive.  The datawriter cycles through the transport instances looking for one that will work.  The datareader must be prepared to establish a connection on any of the transports.

When testing with multiple transport instances, I found the following race condition.  Suppose the transport config contains udp and unreliable multicast and assume best-effort reliability.  (Note that the udp and multicast transport will eventually go away.  However, the problem I'm describing is fundamental to the design of the transport framework.)

After discover completes, the writer will send a message via udp to establish the connection.  The reader will start accepting by calling `accept_datalink` on its transport instances.  The first one will succeed without returning a datalink meaning that it is waiting for something, in this case, it is waiting for the message from the datawriter.  The second transport succeeds and returns a datalink since it is unreliable multicast.  Thus, **the reader assumes that all communication will occur via unreliable multicast**.  Upon receiving the message from the writer, the udp transport responds to the datawriter with an ack and calls `use_datalink` on the reader.  This won't succeed because a datalink has already been established by the unreliable multicast transport.  Upon receiving the ack from the reader, **the writer assumes that all communication will occur via udp**.  At this point, the writer will send data but the reader will drop it and no communication will take place.

As I said, this problem won't happen in the future because the udp and multicast transports will be removed at some point.  However, the problem could easily be replicated using a different set of transports.

A first step may be to require that `accept_datalink` be an asynchronous process that forces proof-of-life.  Taken together, this means that the multicast transport would not be able to return a datalink and would be waiting on some message from the writer to complete the connection.  This creates a scenario where all of the transports are ready to establish connections and racing with each other.

As a side point, there is a similar problem on the writer side.  We would like datalink establishment to based on proof-of-life and therefore asynchronous.  However, this breaks reliability because other pieces of the framework use association as the start of reliability. 

The next issue to address is that of acknowledging the writer.  In the case of udp, the ack is sent immediately.  It could be deferred if it was determined that this was indeed the transport to use.  This pattern would need to be applied to all transports, i.e., they need to cooperate in telling the writer the correct transport to use.  It is an open question whether this could be applied to all transports.

Related to the previous point is the idea that the datawriter may appear to be establishing connections with multple transports due to network delay and message re-ordering.  Thus, the datawriter has the same issue that a transport that they have moved on from may be the one selected by the reader.

Most of these issues come from the design of the multi-transport support:  a datawriter tries transports in order until they find one that works.  This design requires a lot of coordination, handling of corner cases, design limitations on transports, etc.  I think that starting with a different design can side-step many of these problems.  Let all of the transports for the writer attempt to establish connectivity with all of the transports of the reader.  Those that succeed can be prioritized and used for communication.  Sending a message on one transport should be equivalent to sending a message on another transport.  This implies a new division of responsibilities where transports can become simpler and the writer/reader take on more responsibility.
